### ISSUE TITLE
RAM Usage Reduction: Autogenerated UA_DataTypeMember's should be declared const

### ISSUE THREAD

**Issue by wrc-jasonw** [open62541/open62541#1806]
Background: I am on a microcontroller running FreeRTOS. I am currently in the process of trimming RAM usage of my implementation. I figured I may as well share my findings so that others may benefit from the work.

In ua_types_generated.c and ua_transport_generated.c all variables of type "static UA_DataTypeMember" should be declared "static const UA_DataTypeMember" in order to reduce memory usage.

This yielded a "free" savings of 3172 bytes of RAM on the STM32F407. This accounted for about 2.5% of the RAM used by the stack.

As I understand it the datatype arrays of type "UA_DataTypeMember[...]" are constant after the server has started. Therefore they can be declared constant which allows the data to be stored in (relatively cheap) program flash rather than in (precious) RAM.

**Comment by jpfr** [comment]
That's a good contribution!
Can you make a pull request from your improvements?

**Comment by wrc-jasonw** [comment]
Sorry, for the time being I have only modified the auto-generated C files. The real culprit are the python auto-generator scripts.

**Comment by Stasik0** [comment]
@jpfr: cannot same be done for nodes in the nodestore? Since they are immutable anyhow?

**Comment by wrc-jasonw** [comment]
I would be very grateful if that is the case.

Currently (with my very limited understanding of how things work) each node seems to take about 560 bytes. My current task is identifying variables, objects, and datatype that are optional so that I can remove them. so far I have identified 55+ nodes in the default XML that I can remove which should yield a savings of ~31K of RAM.

If nodes can be moved into flash then RAM usage would be reduced ever further.

**Comment by wrc-jasonw** [comment]
Thank you again everybody! This has been very helpful.

Results of my review of the default namespace0:

The XML file provided by the OPC-UA foundation contains many optional objects and datatypes that can be removed to save RAM,

In particular the OPC Binary->Opc.Ua DatatypeDictionary contains ~15kB of XML. In version 1.04 of OPC-UA DatatypeDictionary's were removed from the standard. Also OperationLimits, OperationalLimitsType, ServerDiagnosticSummery, ServerDiagnosticSummeryType and all their children nodes are all optional. The set of standard types like Byte, Float, Image, XmlElement also all can be removed if they are not used by the application.

By doing the above I observed an overall reduction of RAM usage of ~45kB.
